### PR TITLE
BUG: to_dict should return a native datetime object for NumPy backed dataframes

### LINKED ISSUE #21256: Type conversions are skipped in 'to_dict' on single column dataframes
Code to reproduce the error:

```python
import pandas as pd
from datetime import datetime

dfs = {
    'full_df': pd.DataFrame([
        {'int': 1, 'date': datetime.now(), 'str': 'foo', 'float': 1.0, 'bool': True},
    ]),
    'int_df': pd.DataFrame([
        {'int': 1},
    ]),
    'date_df': pd.DataFrame([
        {'date': datetime.now()},
    ]),
    'str_df': pd.DataFrame([
        {'str': 'foo'},
    ]),
    'float_df': pd.DataFrame([
        {'float': 1.0},
    ]),
    'bool_df': pd.DataFrame([
        {'bool': True},
    ])
}

for name, frame in dfs.items():
    print('Types in ' + name)
    for k, v in frame.to_dict('records')[0].items():
        print(type(v))
```
Output:
```python
Types in full_df
<class 'bool'>
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
<class 'float'>
<class 'int'>
<class 'str'>
Types in int_df
<class 'numpy.int64'>
Types in date_df
<class 'pandas._libs.tslibs.timestamps.Timestamp'>
Types in str_df
<class 'str'>
Types in float_df
<class 'numpy.float64'>
Types in bool_df
<class 'numpy.bool_'>
```
#### Problem description

One would expect that the `to_dict()` function returns python native types, or at least does the same to the same type of columns, however it behaves differently as shown above. It seems that type conversion is not invoked when a single column is present in the dataframe.

#### Expected Output

Python native types where it is possible for `int`, `float`, `bool` and `str` types, and if possible, a python `datetime` object instead of `pandas.Timestamp`

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.3.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 142 Stepping 9, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None
LOCALE: None.None

pandas: 0.23.0
pytest: None
pip: 9.0.1
setuptools: 38.2.4
Cython: None
numpy: 1.14.3
scipy: None
pyarrow: None
xarray: None
IPython: 6.3.1
sphinx: 1.7.4
patsy: None
dateutil: 2.7.3
pytz: 2018.4
blosc: None
bottleneck: None
tables: None
numexpr: None
feather: None
matplotlib: None
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 1.0.1
sqlalchemy: 1.2.2
pymysql: None
psycopg2: 2.7.3.2 (dt dec pq3 ext lo64)
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by arw2019**
- [x] closes #21256
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Added a fix for datetime. Added tests for datetime and for bool

### CODE REVIEW

File: pandas/core/common.py [hunk 1]
```diff
@@ -78,17 +78,19 @@ def consensus_name_attr(objs):
     return name
 
 
-def maybe_box_datetimelike(value, dtype=None):
+def maybe_box_datetimelike(value, dtype=None, native=False):
```
Comment (jreback) [inline pandas/core/common.py:81]: I don't like this function at all where it is. move to pandas/core/dtypes/cast.py

call it

maybe_box_to_native

and we don't need a native argument, the results as Timestamps are already correct.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/common.py (hunk 1 above)
Comment (jreback) [reply pandas/core/common.py:81]: likely you would want to add coercsion for int/float/bool as these are numpy dtypes (but do that separately)

Comment (arw2019) [reply pandas/core/common.py]: > I don't like this function at all where it is. move to pandas/core/dtypes/cast.py
> 
> call it
> 
> maybe_box_to_native
> 
> and we don't need a native argument, the results as Timestamps are already correct.

Done both

Comment (arw2019) [reply pandas/core/common.py]: > likely you would want to add coercsion for int/float/bool as these are numpy dtypes (but do that separately)

Yes will do in a follow-on (to resolve #27616)

**Comment by arw2019** [comment]
Ready for re-review (CI  green + addressed comments)

File: pandas/core/common.py [hunk 2]
```diff
@@ -6,19 +6,21 @@
...
 
-from pandas._libs import lib, tslibs
+from pandas._libs import lib
 from pandas._typing import AnyArrayLike, Scalar, T
 from pandas.compat.numpy import np_version_under1p18
 
-from pandas.core.dtypes.cast import construct_1d_object_array_from_listlike
+from pandas.core.dtypes.cast import (
+    construct_1d_object_array_from_listlike,
+    maybe_box_datetimelike,
```
Comment (jreback) [inline pandas/core/common.py:22]: why don't you move dict_compat *also* to pandas/core/dtypes/cast.py  (which uses maybe_box_datetimelike)

File: pandas/core/dtypes/cast.py [hunk 3]
```diff
@@ -134,6 +134,21 @@ def is_nested_object(obj) -> bool:
     return False
 
 
+def maybe_box_datetimelike(value, dtype=None):
+    # turn a datetime like into a Timestamp/timedelta as needed
```
Comment (jreback) [inline pandas/core/dtypes/cast.py:138]: can you add a full-docstring, type if you can

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by arw2019** [comment]
Ready for re-review. Addressed comments + CI green

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @arw2019 very nice

**Pull request merged**